File: src/c/auxlib/shade.c

/* *** LAST REVISED ON  7-JAN-1994 17:13:41.80 */
/* *** SOURCE FILE: [LONGLIB93.SOURCES.C.AUXLIB]SHADE.C */

#include "my.h"

/* Table of constant values */

static integer c__1 = 1;

/* Subroutine */ int shade_(a, b, n, i, l, d, t, w, ma, am, da, bm, db)
real *a, *b;
integer *n, *i, *l;
real *d, *t, *w;
integer *ma;
real *am, *da, *bm, *db;
{
    /* Initialized data */

    integer ku = 3;
    integer kd = 2;
    integer kp[2] = {2,3};

    /* System generated locals */
    integer i__1;
    real r__1, r__2;

    /* Builtin functions */
    double sin(), cos();

    /* External functions */
    extern /* Subroutine */ int newpen_();
    extern /* Subroutine */ int plot_();

    /* Local variables */
    static integer i0, i1, i2, i3;
    static real w1, w2, x1, ai, bi;
    static real ao, bo;
    static integer kc;
    static real ct, dt;
    static real th, yc;
    static integer lt;
    static real yl, st, ys, xt, yt;
    static real tx0, ty0, ty1, yt0, yt1, xt0, xt1;

/* 	SHADES AN AREA DEFINED BY THE A AND B ARRAYS */

/* A    (R) ARRAY CONTAINING ABSCISSA VALUES */
/* B    (R) ARRAY CONTAINING ORDINATE VALUES */
/* N    (I) NUMBER OF POINTS IN ARRAY */
/* I    (I)INCREMENT BETWEEN POINTS IN ARRAY */
/* L    (I) SHADE FORMAT CONTROL */
/* 		-3 CLEAR AREA AND OUTLINE */
/* 		-2 CLEAR AREA */
/* 		-1 CLEAR OUTLINE */
/* 		 0 NO ACTION */
/* 		+1 SHADE OUTLINE */
/* 		+2 SHADE AREA */
/* 		+3 SHADE AREA AND OUTLINE */
/* D    (D) DISTANCE BETWEEN SHADING LINES */
/* T    (D) ANGLE OF SHADING LINES */
/* W    (D) WORKING ARRAY DIMENSIONED AT LEAST (3*N) */
/* MA   (I) TYPE OF SHADING (LINE TYPE) IF MA > 0 */
/* AM   (R) MINIMUM VALUE FOR A ARRAY */
/* DA   (R) SCALE FACTOR FOR A ARRAY */
/* BM   (R) MINIMUM VALUE FOR B ARRAY */
/* DB   (R) SCALE FACTOR FOR B ARRAY */

/* 	EQUIVALENCES ARE FOR CONVENIENCE ONLY */

    /* Parameter adjustments */
    w -= 4;
    --b;
    --a;

    /* Function Body */

    if (*n <= 2) {
	goto L110;
    }
    if (*i <= 0) {
	goto L110;
    }
    ao = *am;
    ai = *da;
    bo = *bm;
    bi = *db;
    if (ai <= (float)0.) {
	ai = (float)1.;
    }
    if (bi <= (float)0.) {
	bi = (float)1.;
    }
    lt = abs(*l);
    kd = 2;
    if (*l < 0) {
	kd = 9;
	kp[1] = 9;
    }
    if (lt <= 1) {
	goto L90;
    }
    dt = *d;
    if (dt <= (float)0.) {
	dt = (float).005;
    }
    r__1 = *t / (float)360.;
    th = *t - (int) (r__1) * (float)360.;
    if (th < (float)0.) {
	th += (float)360.;
    }
    th *= (float).0174532;
    st = sin(th);
    ct = cos(th);
    r__1 = (b[1] - bo) / bi;
    r__2 = -(doublereal)((a[1] - ao) / ai);
    ys = r__1 * ct + r__2 * st;
    yl = ys;
    i1 = 1;
    i__1 = *n;
    for (i0 = 1; i0 <= i__1; ++i0) {
	tx0 = (a[i1] - ao) / ai;
	ty0 = (b[i1] - bo) / bi;
	w[i0 * 3 + 1] = tx0 * ct + ty0 * st;
	r__1 = -(doublereal)tx0;
	ty1 = ty0 * ct + r__1 * st;
	w[i0 * 3 + 2] = ty1;
	ys = dmin(ys,ty1);
	yl = dmax(yl,ty1);
	i1 += *i;
/* L20: */
    }
    yc = ys + dt * (float).5;
    if ((real) (*ma) >= 0) {
	newpen_(ma);
    }
L30:
    if (yc > yl) {
	goto L90;
    }
    i1 = 0;
    i__1 = *n;
    for (i0 = 1; i0 <= i__1; ++i0) {
	i2 = i0 + 1;
	if (i2 > *n) {
	    i2 = 1;
	}
	yt0 = w[i0 * 3 + 2];
	yt1 = w[i2 * 3 + 2];
	w1 = dmin(yt0,yt1);
	if (yc < w1) {
	    goto L60;
	}
	w2 = dmax(yt0,yt1);
	if (yc >= w2) {
	    goto L60;
	}
	xt0 = w[i0 * 3 + 1];
	xt1 = w[i2 * 3 + 1];
	x1 = xt0 - (yt0 - yc) * (xt1 - xt0) / (yt1 - yt0);
	++i1;
	i2 = i1;
L40:
	if (i2 == 1) {
	    goto L50;
	}
	i3 = i2 - 1;
	if (w[i3 * 3 + 3] >= x1) {
	    goto L50;
	}
	w[i2 * 3 + 3] = w[i3 * 3 + 3];
	i2 = i3;
	goto L40;
L50:
	w[i2 * 3 + 3] = x1;
L60:
	;
    }
    if (i1 <= 0) {
	goto L80;
    }
    i2 = 1;
    i__1 = i1;
    for (i0 = 1; i0 <= i__1; ++i0) {
	xt0 = w[i0 * 3 + 3];
	r__1 = -(doublereal)yc;
	xt = xt0 * ct + r__1 * st;
	yt = yc * ct + xt0 * st;
	i2 = 3 - i2;
	kc = kp[i2 - 1];
	plot_(&xt, &yt, &kc);
/* L70: */
    }
L80:
    yc += dt;
    goto L30;
L90:
    if (lt / 2 << 1 == lt) {
	goto L110;
    }
    r__1 = (a[1] - ao) / ai;
    r__2 = (b[1] - bo) / bi;
    plot_(&r__1, &r__2, &ku);
    i1 = 1;
    i__1 = *n;
    for (i0 = 2; i0 <= i__1; ++i0) {
	i1 += *i;
	r__1 = (a[i1] - ao) / ai;
	r__2 = (b[i1] - bo) / bi;
	plot_(&r__1, &r__2, &kd);
/* L100: */
    }
    r__1 = (a[1] - ao) / ai;
    r__2 = (b[1] - bo) / bi;
    plot_(&r__1, &r__2, &kd);

/* 	RESET PEN TYPE */

L110:
    if ((real) (*ma) > 0) {
	newpen_(&c__1);
    }
    return 0;
} /* shade_ */

File: src/f/auxlib/shade.f
C
C
	SUBROUTINE SHADE(A,B,N,I,L,D,T,W,MA,AM,DA,BM,DB)
C
C	SHADES AN AREA DEFINED BY THE A AND B ARRAYS
C
C A    (R) ARRAY CONTAINING ABSCISSA VALUES
C B    (R) ARRAY CONTAINING ORDINATE VALUES
C N    (I) NUMBER OF POINTS IN ARRAY
C I    (I)INCREMENT BETWEEN POINTS IN ARRAY
C L    (I) SHADE FORMAT CONTROL
C		-3 CLEAR AREA AND OUTLINE
C		-2 CLEAR AREA
C		-1 CLEAR OUTLINE
C		 0 NO ACTION
C		+1 SHADE OUTLINE
C		+2 SHADE AREA
C		+3 SHADE AREA AND OUTLINE
C D    (D) DISTANCE BETWEEN SHADING LINES
C T    (D) ANGLE OF SHADING LINES
C W    (D) WORKING ARRAY DIMENSIONED AT LEAST (3*N)
C MA   (I) TYPE OF SHADING (LINE TYPE) IF MA > 0
C AM   (R) MINIMUM VALUE FOR A ARRAY
C DA   (R) SCALE FACTOR FOR A ARRAY
C BM   (R) MINIMUM VALUE FOR B ARRAY
C DB   (R) SCALE FACTOR FOR B ARRAY
C
	INTEGER KP(2)
	REAL A(1),B(1),W(3,1)
C
C	EQUIVALENCES ARE FOR CONVENIENCE ONLY
C
	EQUIVALENCE (KP(1),KD),(KP(2),KU)
	DATA KP/2,3/
C
	ROT(X,Y)=X*CT+Y*ST
	SCL(Q,R,S)=(Q-R)/S
C
	IF (N.LE.2) GOTO 110
	IF (I.LE.0) GOTO 110
	AO=AM
	AI=DA
	BO=BM
	BI=DB
	IF (AI.LE.0.0) AI=1.0
	IF (BI.LE.0.0) BI=1.0
	LT=IABS(L)
	KD=2
	IF (L.LT.0) KD=9
	IF (LT.LE.1) GOTO 90
	DT=D
	IF (DT.LE.0.0) DT=0.005
	TH=T-(AINT(T/360.0))*360.0
	IF (TH.LT.0.0) TH=TH+360.0
	TH=TH*0.0174532
	ST=SIN(TH)
	CT=COS(TH)
	YS=ROT(SCL(B(1),BO,BI),-SCL(A(1),AO,AI))
	YL=YS
	I1=1
	DO 20 I0=1,N
		TX0=SCL(A(I1),AO,AI)
		TY0=SCL(B(I1),BO,BI)
		W(1,I0)=ROT(TX0,TY0)
		TY1=ROT(TY0,-TX0)
		W(2,I0)=TY1
		YS=AMIN1(YS,TY1)
		YL=AMAX1(YL,TY1)
		I1=I1+I
20	CONTINUE
	YC=YS+0.5*DT
	IF (MA.GE.0) CALL NEWPEN(MA)
30	IF (YC.GT.YL) GOTO 90
	I1=0
	DO 60 I0=1,N
		I2=I0+1
		IF(I2.GT.N)I2=1
		YT0=W(2,I0)
		YT1=W(2,I2)
		W1=AMIN1(YT0,YT1)
		IF(YC.LT.W1)GOTO 60
		W2=AMAX1(YT0,YT1)
		IF (YC.GE.W2) GOTO 60
		XT0=W(1,I0)
		XT1=W(1,I2)
		X1=XT0-(YT0-YC)*(XT1-XT0)/(YT1-YT0)
		I1=I1+1
		I2=I1
40		IF (I2.EQ.1) GOTO 50
		I3=I2-1
		IF (W(3,I3).GE.X1) GOTO 50
		W(3,I2)=W(3,I3)
		I2=I3
		GOTO 40
50		W(3,I2)=X1
60	CONTINUE
	IF (I1.LE.0) GOTO 80
	I2=1
	DO 70 I0=1,I1
		XT0=W(3,I0)
		XT=ROT(XT0,-YC)
		YT=ROT(YC,XT0)
		I2=3-I2
		KC=KP(I2)
		CALL PLOT(XT,YT,KC)
70	CONTINUE
80	YC=YC+DT
	GOTO 30
90	IF (2*(LT/2).EQ.LT) GOTO 110
	CALL PLOT(SCL(A(1),AO,AI),SCL(B(1),BO,BI),KU)
	I1=1
	DO 100 I0=2,N
		I1=I1+I
		CALL PLOT(SCL(A(I1),AO,AI),SCL(B(I1),BO,BI),KD)
100	CONTINUE
	CALL PLOT(SCL(A(1),AO,AI),SCL(B(1),BO,BI),KD)
C
C	RESET PEN TYPE
C
110	IF (MA.GT.0) CALL NEWPEN(1)
	RETURN
	END
